[MLIR] AMDGPUToROCDL: Use a bitcast op to reintepret a vector of i8 as single integer.

bjacob opened the pull request:
Found by inspecting AMDGPU assembly - so the arithmetic ops created there were definitely making their way into the target ISA. A `LLVM::BitcastOp` seems equivalent, and evaporates as expected in the target asm.

Along the way, I thought that this helper function `mfmaConcatIfNeeded` could be renamed to `convertMFMAVectorOperand` to better convey its contract;  so I don't need to think about whether a bitcast is a legitimate "concat" :-)

bjacob committed:
bitcast

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>

arsenm reviewed (commented):
missing test changes

bjacob commented:
> missing test changes

It's because the test actually only checks that the `mfma` op is as expected, it wasn't covering the IR that is being simplified here. I will add that.

kuhar reviewed (commented):
Can we have a testcase for this?

kuhar commented:
Ah, just saw that @arsenm had the same comment

krzysz00 reviewed (approved):
Approved conditional on "whoops, I forgot to add a test for that bit, please fix" above

bjacob committed:
add test

bjacob commented:
Added the test.  Note: added `-cse` to the existing test file, because this test passes the same Value for LHS and RHS, so we are getting the same IR generated twice; writing the CHECKs against the non-CSE'd form seems like it would be fragile.

bjacob committed:
Also check for bitcast on bf8/fp8 testcases

arsenm commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -351,39 +351,23 @@ struct SchedBarrierOpLowering : public ConvertOpToLLVMPattern<SchedBarrierOp> {
-    for (int64_t i = 0; i < numBytes; ++i) {
-      Value idxConst = createI32Constant(rewriter, loc, i);
-      Value element =
-          rewriter.create<LLVM::ExtractElementOp>(loc, input, idxConst);
-      Value extended = rewriter.create<LLVM::ZExtOp>(loc, destType, element);
-      Value shiftConst = rewriter.create<LLVM::ConstantOp>(
-          loc, destType, rewriter.getIntegerAttr(destType, i * 8));
-      Value shifted = rewriter.create<LLVM::ShlOp>(loc, extended, shiftConst);
-      result = rewriter.create<LLVM::OrOp>(loc, result, shifted);
+    if (vectorType.getElementType().isInteger(8)) {
+      return rewriter.create<LLVM::BitcastOp>(
+          loc, rewriter.getIntegerType(vectorType.getNumElements() * 8), input);
     }
-    return result;
   }
   return input;
```
This whole function can just create one bitcast? I don't see why you need to consider the element types. Especially since bf16 should be natively consumed now

bjacob replied:
I thought about that, but thought that maybe we want to avoid creating identity bitcasts unconditionally. And then if we're doing it conditionally when it would be needed, the code starts looking like its current form.
Feel free to improve this code further in a follow-up!

arsenm replied:
Avoiding identity bitcasts is one compare that the type matches (and the IRBuilder at least does that for you)

bjacob replied:
Good point! But I mean: we sometimes want to bitcast, and sometimes not. We don't bitcast f32 and f16's. So (even after the bf16 simplification you mentioned) we still need to have some logic based on element types.  I'd rather defer any further simplification to you as a follow-up, since you were aware of things such as this bf16 simplification, which I wasn't.

bjacob replied:
There's also a nuance: in the `bf16` case, we bitcast to a *vector* of `i16`, while in the case of byte-size element types, we bitcast to a raw *integer* of higher bit width, not a vector of integers of the original element type bit width.

krzysz00 replied:
Also, the MFMA intrinsics still take `i16`, last I checked
